### ISSUE TITLE
not correctly joining lines for HTML

### ISSUE THREAD

**Issue by garretwilson** [beautify-web/js-beautify#1033]
This is really hurting me. Here is an example HTML file:

``` html
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html lang="en-US" xmlns="http://www.w3.org/1999/xhtml">

<head>
    <title>Test Line Join</title>
</head>

<body>
    <p>The key words <q><span class="spec-must">must</span></q>, <q><span

          class="spec-must-not">must not</span></q>, <q><span class="spec-must">required</span></q>, <q><span

          class="spec-must">shall</span></q>, <q><span class="spec-must-not">shall not</span></q>, <q><span

          class="spec-should">should</span></q>, <q><span class="spec-should-not">should not</span></q>, <q><span

          class="spec-should">recommended</span></q>, <q><span class="spec-may">may</span></q>, and <q><span

          class="spec-may">optional</span></q> in this document are to be interpreted as described in <a href="#ref-rfc2119" class="ref">RFC 2119</a>.</p>
</body>

</html>
```

js-beautify should be getting rid of all that whitespace inside the `<span …>` tag. I want js-beautify to give me this:

``` html
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html>
<html lang="en-US" xmlns="http://www.w3.org/1999/xhtml">

<head>
    <title>Test Line Join</title>
</head>

<body>
    <p>The key words <q><span class="spec-must">must</span></q>, <q><span class="spec-must-not">must not</span></q>, <q><span class="spec-must">required</span></q>, <q><span class="spec-must">shall</span></q>, <q><span class="spec-must-not">shall not</span></q>, <q><span class="spec-should">should</span></q>, <q><span class="spec-should-not">should not</span></q>, <q><span class="spec-should">recommended</span></q>, <q><span class="spec-may">may</span></q>, and <q><span class="spec-may">optional</span></q> in this document are to be interpreted as described in <a href="#ref-rfc2119" class="ref">RFC 2119</a>.</p>
</body>

</html>
```

But it doesn't. This is a huge problem and is really dragging down the workflow. (The reason the lines are like that to begin with is a bug in BlueGriffon. So bugs in js-beautify are preventing it from fixing the bugs in BlueGriffon. Such is life.)

See #841.

Perhaps you could point me to a place in the code I might go to try to put in some workaround?

**Comment by HookyQR** [comment]
`q` tags are not formatted by default. If you add an `"unformated"` option to your `.jsbeautifyrc` file, with the defaults minus the `q` you'll get each of the `q` groups on their own line at least, with the `span`s correctly un-whitespaced. 

eg:
```javascript
"unformatted": ["a", "abbr", "area", "audio", "b", "bdi", "bdo", "br", "button", "canvas", "cite", "code", "data",
		"datalist", "del", "dfn", "em", "embed", "i", "iframe", "img", "input", "ins", "kbd", "keygen", "label", "map",
		"mark", "math", "meter", "noscript", "object", "output", "progress", "ruby", "s", "samp", "select", "small", "span",
		"strong", "sub", "sup", "svg", "template", "textarea", "time", "u", "var", "video", "wbr", "text", "acronym",
		"address", "big", "dt", "ins", "strike", "tt"],
```

**Comment by garretwilson** [comment]
Yes, but you see @HookyQR , I _want_ the '<q>' tags formatted --- I just want them _inline_, not block. js-beautify conflates the two concepts. See #841, as I mentioned above.

@bitwiseman I want to help and not just complain. As I mentioned above, could you point me to where I might look in the code to track this down? Or do we have to just rewrite it all from scratch?

**Comment by bitwiseman** [comment]
@garretwilson - just rewrite it all from scratch.  👍  Sooner or later, the HTML and CSS beautifiers will need to be rewritten to at least have a parse stage like I did with the JS beautifier.  It's just a major undertaking.

**Comment by garretwilson** [comment]
Yes, but I'm extremely, extremely overloaded with work --- and every time I add a tool to help me with my work, it seems to have a huge bug that adds _more_ to my workload. (As an example, the reason I need this capability in the first place is that BlueGriffon has a stupid bug that the owner ignores by making line breaks even when we request in the configuration never to break lines, and so I have to use js-beautify to fix that, and _it_ has bugs that doesn't join lines... and now Atom introduced a workaround where it is automatically soft-wrapping lines (because it crashes on what it considers long lines) --- one after another, more bugs in the tools to fix bugs.

So @bitwiseman , would you or anyone else be interested in fixing this bug or at least creating a workaround in the code, for payment? In general js-beautify for XHTML is hanging together, and if we could fix a couple of these big issues it would really help. How much would you charge to do something about this? Please contact me (you or anyone else) offline. Thanks.

**Comment by pubkey** [comment]
I'm trying to fix the same problem for the last 3 days.
My question still is why beautify does not rely on the indent-character which is defined in my eslint.json?
Shouldn't the project-based configuration overwrite every other setting?

**Comment by bitwiseman** [comment]
@garretwilson 
Thanks for the offer of compensation.  I appreciate the thought.  For me, much like @Glavin001,  my main limitation is not money, but time.

What I need is people who can write code and tests that I don't have to spend cycles debugging or fixing to join this project and contribute fixes.  Until that happens, issues will continue to be addressed as I find time (that is to say, very slowly).

**Comment by garretwilson** [comment]
OK, so I'm trying out this Bountysource thing. While I want to eventually address js-beautify and atom-beautify problems, I'm going directly to the ultimate source of some of pain, and then work my way up if there is any success.

https://www.bountysource.com/issues/29803090-line-wrapping-setting-ignored

1. Basically the whole thing starts because the BlueGriffon editor has the blatant bug https://github.com/therealglazou/bluegriffon/issues/7 that arbitrarily adds line breaks in the middle of elements even though line wrapping are turned off. (To top it off, it adds them as `LF` even on the Windows platform, where it uses `CRLF` on the lines I _really_ want: https://github.com/therealglazou/bluegriffon/issues/5. And you'll remember that js-beautify at one time would choose the wrong line ending: https://github.com/beautify-web/js-beautify/issues/829. So there were bugs upon bugs upon bugs. Thank goodness https://github.com/beautify-web/js-beautify/issues/829 is fixed. Thank you very much)

2. Because of https://github.com/beautify-web/js-beautify/issues/841 , atom-beautify refuses to re-join the lines that should have never been broken in the first place.

So when one tool's bugs prevent working around another tool's bugs, it leaves me frustrated and a bit grouchy. Cross your fingers that this Bountysource thing works, and if so I can go up the chain start knocking off some of the nasty js-beautify ones as well, such as this one.

**Comment by garretwilson** [comment]
@bitwiseman , I've thought about this a lot, and I think I have a way forward with this issue. I would like to put up a bounty on Bountysource for this bug, but before I put up the money I want to get your approval for the solution so I can know it would be integrated.

The following solution is based on the understanding that the current logic must add newlines in two separate cases:

* Whenever it thinks that a line should be wrapped (if wrapping is turned on). I'll refer to this as a "wrap-newline".
* Whenever it is ready to start a new element that is not in `unformatted`. I'll refer to this as a "block-newline".

Here is what I propose:

1. Create a new setting `inline` with the exact same value as `unformatted`. (These are the HTML5 inline elements you and I discussed and which you fixed in #840.)
2. Find the line(s) in which the current formatting logic adds a block-newline. (All the code for wrap-newline will go untouched.) For every block-newline occurrence, add the following logic: "if the new element (to appear after the block-newline) is in `inline`, do not add the block-newline."

I have _not_ looked at the code, but seriously, that seems pretty simple and straightforward to me, regardless of the complexity of the code. There has to be some line(s) at which you perform a block-newline break, and I don't see why one can't just do a check to see if the new element is in the `newline` list or not. (I could be completely wrong about the feasibility of this solution, but that would have to be some really jumbled code not to allow this approach.)

With the above approach, if no changes are made to a user's configuration, the formatting _will function 100% exactly as it did before_. However for those of us who realize there is a distinction between "inline vs block" and "formatted vs unformatted", we'll merely set our `unformatted` setting to `""` (that is, we don't want any elements to be "unformatted"). **Suddenly all of our `<a>`, `<span>`, `<q>`, etc. elements will be formatted correctly, yet without them being placed on separately lines as if they were block elements!**

@bitwiseman I'd like to get your approval that you would integrate such a solution if I were to write it or pay someone else to do it.

Thank you! I think we all want to get this fixed, and I'm trying the best I can to find a way to help get it done.

**Comment by bitwiseman** [comment]
@garretwilson - 

Either way, I'd rather see the behavior change and be consistent for all users.  I've never argued against the correctness of what you described in #841, only that I didn't have bandwidth to do it.  

I'd like to make sure you understand, the bounty poster does not get any say in how a fix is designed or implemented.  The bounty is paid when the issue is fixed.  Period.  I appreciate you proposing a design and asking for a review, whoever submits a fix is not required to pay any attention to that proposal.

Are you planning to bounty this issue or #841?

**Comment by garretwilson** [comment]
> Either way, I'd rather see the behavior change and be consistent for all users.

Sure, I'd like that even better! I just wasn't sure you wanted to make such a change that would affect some behavior without increasing the major version number.

We could therefore set `unformatted` to `""`, and move all the current `unformatted` elements to `inline`. That would make all the inline element format correctly, but still be inline; and by default no elements would be "unformatted".

If people considered this issue a bug, then everyone would be happy. The only problem would occur for people who both 1) thought "unformatted" really meant "unformatted", and 2) did not change the value of "unformatted". They would then be surprised when js-beautify started formatting their elements (while still leaving them inline). But I would imagine that if people _are_ using `unformatted`, they have explicitly specified which elements they want by overriding `unformatted`, so they would never see any difference.

The only reason I suggested leaving `unformatted` the way it was is that I was afraid you wouldn't want existing behavior changed, and I was trying to make it more palatable for you.

> The bounty is paid when the issue is fixed. Period.

If they want to rewrite the thing, that's fine with me. I was just suggesting an ultra-easy approach to getting it done and making some money. I just wanted to make sure you were fine with it.

> Are you planning to bounty this issue or #841?

It makes no difference to me --- it may be that they are really the same thing: #841 explains the design problem, while this issue #1033 illustrates a negative behavior resulting from the design problem. Do you have a preference on which issue I put the bounty on? And more importantly, are you OK with what I proposed? (Because I imagine the bounty person will choose the easiest path, and I don't think there could be an approach much easier than what I laid out above.)

(If this discussion goes on much longer, I may just do it myself...)

**Comment by garretwilson** [comment]
> Do you have a preference on which issue I put the bounty on? And more importantly, are you OK with what I proposed?

Did you have an opinion here? I'll be putting up actual dough, so it would be nice to know ahead of time if you're OK with this direction...

**Comment by garretwilson** [comment]
Never mind; it's easier just to start the thing.

https://www.bountysource.com/orders/68866

**Comment by garretwilson** [comment]
I just realized that I had [posted](https://github.com/beautify-web/js-beautify/issues/1033#issuecomment-303618771) the incorrect bounty link a year ago. I updated it. I see someone says they are working on it. Please let me know any questions you have or clarifications you need. Thanks.

**Comment by madman-bob** [comment]
Thanks. I'm replacing `unformatted` tags with `inline` tags, the intent of `unformatted` seeming to be contained in `content_unformatted`.

It's not a question, but may I thank you for the large number of tests. It makes me confident that I'm not breaking anything with my changes.

**Comment by garretwilson** [comment]
@madman-bob , you don't know what a huge help this will be do me. Anything you need on my end, just ask.

**Comment by garretwilson** [comment]
And just in case you missed it, @madman-bob , #841 has a more in-depth discussion of the semantics of this bug. I recommend reading that ticket as well, if you haven't already. Thanks!!

**Comment by madman-bob** [comment]
Ready for review: https://github.com/beautify-web/js-beautify/pull/1407

**Comment by bitwiseman** [comment]
@garretwilson 
This is the branch for that PR: https://github.com/madman-bob/js-beautify/tree/format-inline-tags
It includes the built files under https://github.com/madman-bob/js-beautify/tree/format-inline-tags/js/lib

**Comment by bitwiseman** [comment]
@garretwilson 
It is important to me that you have a chance to test this before I merge this.  
I don't have an npm package for you, so you'll need to pull the files linked above and copy them into the folder where you have the beautifier installed.  Sorry for the hacky process, but it is what I have right now.

**Comment by garretwilson** [comment]
I apologize for not getting to this yet. I'm being overloaded with a couple of very stressful things right now. I'll try my best to get to this by this weekend.

So if I pull the new files, do you think I can simply replace the ones in the atom-beautify plugin? If so, that would make this really simple to test.

**Comment by garretwilson** [comment]
Oh, this is something cool I found in one file. BlueGriffon would produce something like this (with preformatted content removed for brevity:

```html
    <aside class="note far half">The Super POM already contains the following definitions, which is why the standard resource directories are copied by default:
      <pre class="line-numbers"><code class="language-xml">
…
</code></pre></aside>
```

Before now, js-beautify would live this untouched except for tabifying the non-`<pre>` indentions, as per my settings. (The `</code>` is correctly not indented, because it's inside a `<pre>`.) But now with the new changes, we get this:

```html5
	<aside class="note far half">The Super POM already contains the following definitions, which is why the standard resource directories are copied by default:
		<pre class="line-numbers"><code class="language-xml">
…
</code></pre>
	</aside>
```

Note that the trailing `</aside>` is now indented! I hadn't noticed this before, as the biggest problem earlier was the newlines inside phrasing/inline elements. But as `</aside>` should be block formatted, it should indeed come on another line and be indented. So it appears there have been some possibly unrelated positive side effects from these changes as well. 👍

**Comment by garretwilson** [comment]
I'm super-excited about this issue being fixed! I just came in for the night. Remind me how to mark the bounty as done so that @madman-bob can be awarded as soon as possible. I just read the FAQ and I think maybe I'll receive some email for me to confirm.

**Comment by bitwiseman** [comment]
This if available in [1.8.0-rc2](https://github.com/beautify-web/js-beautify/milestone/34?closed=1).

**Comment by garretwilson** [comment]
@madman-bob, @bitwiseman, and others, I cannot begin to express my appreciation for finally getting this bug fixed and released. After a little prodding on the atom-beautify side, we finally got this into the Atom plugin as well.

It's hard to describe just how refreshing it is, when I reformat the ugly, buggy HTML that BlueGriffon puts out, to not have to manually search for those weird, wrapped lines that BlueGriffon should not have put in. Now js-beautify just fixes them, automatically (like it always should have). It's so amazing that at times I think I'm dreaming. I cannot count up all the time this is saving me, reducing my workflow, and also reducing my irritation level (or at least not adding to it 😉 ).

Thank you, thank you, thank you.

And after I finally gave in and tried Visual Studio Code (after a long list of stupid Atom problems, the last one being that it reset my entire configuration for no reason one day; see https://github.com/atom/atom/issues/18229), I found out that there is a [Beautify plugin for VS Code](https://marketplace.visualstudio.com/items?itemName=HookyQR.beautifyl)!

https://github.com/HookyQR/VSCodeBeautify

I was almost afraid to check, but yes… _yes_… the plugin [includes the latest js-beautify 1.8.x](https://marketplace.visualstudio.com/items/HookyQR.beautify/changelog)!! This means that after installing VSCodeBeautify, I can open up my HTML and format it _exactly like I did with atom-beautify_ (except using a different shortcut key), with the results being exactly the same—and it even uses the exact same `.jsbeautifyrc` file! This is the smoothest transition I could never have even dreamed possible. For once—_for once_—everything worked just like it should have.

I am very grateful. Thanks for making this happen.
